## Data Cleaning Process
### Overview
The data cleaning process involves two datasets related to California wildfire incidents from 2013 to the present. The goal is to prepare these datasets for analysis by removing unnecessary columns, handling missing or duplicate values, and merging the data effectively.
### Datasets Overview:
- Dataset One: 'mapdataall.csv'
    - Source: California Department of Forestry and Fire Protection
    - Data: Wildfire incidents in California from 2013 to the present.
    - Shape: (2835, 23)
    - Columns Retained: 9 columns selected for analysis.
      
- Dataset Two: 'California_Fire_Incidents.csv'
    - Source: Kaggle - California Wildfire Incidents
    - Data: Wildfire incidents in California from 2013 to 2020.
    - Shape: (1636, 40)
    - Columns Retained: 14 columns selected for analysis.

## California fire Dataset One

In [3]:
import pandas as pd 
import numpy as np

In [4]:
#Import dataset one
CA_fire_df = pd.read_csv('./Data source/mapdataall.csv')

In [5]:
#Display head
CA_fire_df.head()

incident_name incident_is_final incident_date_last_update  \
0    Bridge Fire                 Y      2018-01-09T13:46:00Z   
1      Pala Fire                 Y      2020-09-16T14:07:35Z   
2     River Fire                 Y      2022-10-24T11:39:23Z   
3  Fawnskin Fire                 Y      2013-04-22T09:00:00Z   
4      Gold Fire                 Y      2013-05-01T07:00:00Z   

  incident_date_created          incident_administrative_unit  \
0  2017-10-31T11:22:00Z       Shasta-Trinity National Forest    
1  2009-05-24T14:56:00Z               CAL FIRE San Diego Unit   
2  2013-02-24T08:16:00Z          CAL FIRE San Bernardino Unit   
3  2013-04-20T17:30:00Z        San Bernardino National Forest   
4  2013-04-30T12:59:00Z  CAL FIRE Madera-Mariposa-Merced Unit   

   incident_administrative_unit_url incident_county  \
0                               NaN          Shasta   
1                               NaN       San Diego   
2                               NaN            Inyo   
3                               NaN  San Bernardino   
4                               NaN          Madera   

                                   incident_location  incident_acres_burned  \
0  I-5 and Turntable Bay, 7 miles NE of Shasta Lake                    37.0   
1        Hwy 76 and Pala Temecula, northwest of Pala                  122.0   
2  south of Narrow Gauge Rd & north of Hwy 136, e...                  407.0   
3  west of Delamar Mountain, north of the communi...                   30.0   
4  Between Road 210 and Road 200 near Fine Gold C...                  274.0   

   incident_containment  ... incident_latitude incident_type  \
0                 100.0  ...         40.774000           NaN   
1                 100.0  ...          1.000000      Wildfire   
2                 100.0  ...         36.602575           NaN   
3                 100.0  ...         34.288877           NaN   
4                 100.0  ...         37.116295           NaN   

                            incident_id  \
0  2ca11d45-8139-4c16-8af0-880d99b21e82   
1  8f61f461-552d-4538-b186-35ab030da416   
2  094719ba-a47b-4abb-9ec5-a506b2b9fd23   
3  58f89ff8-bd3e-4355-b1c0-8fa05c747d3f   
4  357ffc13-bef9-48eb-810f-c5de851972eb   

                                        incident_url  \
0  https://www.fire.ca.gov/incidents/2017/10/31/b...   
1  https://www.fire.ca.gov/incidents/2009/5/24/pa...   
2  https://www.fire.ca.gov/incidents/2013/2/24/ri...   
3  https://www.fire.ca.gov/incidents/2013/4/20/fa...   
4  https://www.fire.ca.gov/incidents/2013/4/30/go...   

  incident_date_extinguished incident_dateonly_extinguished  \
0       2018-01-09T13:46:00Z                     2018-01-09   
1       2009-05-25T00:00:00Z                     2009-05-25   
2       2013-02-28T20:00:00Z                     2013-02-28   
3       2013-04-22T09:00:00Z                     2013-04-22   
4       2013-05-01T07:00:00Z                     2013-05-01   

  incident_dateonly_created is_active calfire_incident notification_desired  
0                2017-10-31         N            False                False  
1                2009-05-24         N             True                False  
2                2013-02-24         N             True                False  
3                2013-04-20         N            False                False  
4                2013-04-30         N             True                False  

[5 rows x 23 columns]

In [6]:
CA_fire_df.shape

(2835, 23)

In [7]:
CA_fire_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2835 entries, 0 to 2834
Data columns (total 23 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   incident_name                     2835 non-null   object 
 1   incident_is_final                 2835 non-null   object 
 2   incident_date_last_update         2835 non-null   object 
 3   incident_date_created             2835 non-null   object 
 4   incident_administrative_unit      2779 non-null   object 
 5   incident_administrative_unit_url  0 non-null      float64
 6   incident_county                   2825 non-null   object 
 7   incident_location                 2835 non-null   object 
 8   incident_acres_burned             2782 non-null   float64
 9   incident_containment              2781 non-null   float64
 10  incident_control                  117 non-null    object 
 11  incident_cooperating_agencies     1659 non-null   object 
 12  incide

In [8]:
CA_fire_df.columns

Index(['incident_name', 'incident_is_final', 'incident_date_last_update',
       'incident_date_created', 'incident_administrative_unit',
       'incident_administrative_unit_url', 'incident_county',
       'incident_location', 'incident_acres_burned', 'incident_containment',
       'incident_control', 'incident_cooperating_agencies',
       'incident_longitude', 'incident_latitude', 'incident_type',
       'incident_id', 'incident_url', 'incident_date_extinguished',
       'incident_dateonly_extinguished', 'incident_dateonly_created',
       'is_active', 'calfire_incident', 'notification_desired'],
      dtype='object')

In [9]:
CA_fire_df = CA_fire_df.drop(columns=['incident_is_final','incident_date_last_update','incident_administrative_unit_url',
                                    'incident_location', 'incident_containment','incident_control','incident_cooperating_agencies',
                                     'incident_type', 'incident_url','incident_dateonly_extinguished', 'incident_dateonly_created',
                                     'is_active', 'calfire_incident', 'notification_desired'])

In [10]:
CA_fire_df.head(5)

incident_name incident_date_created          incident_administrative_unit  \
0    Bridge Fire  2017-10-31T11:22:00Z       Shasta-Trinity National Forest    
1      Pala Fire  2009-05-24T14:56:00Z               CAL FIRE San Diego Unit   
2     River Fire  2013-02-24T08:16:00Z          CAL FIRE San Bernardino Unit   
3  Fawnskin Fire  2013-04-20T17:30:00Z        San Bernardino National Forest   
4      Gold Fire  2013-04-30T12:59:00Z  CAL FIRE Madera-Mariposa-Merced Unit   

  incident_county  incident_acres_burned  incident_longitude  \
0          Shasta                   37.0         -122.309000   
1       San Diego                  122.0            1.000000   
2            Inyo                  407.0         -118.016510   
3  San Bernardino                   30.0         -116.941311   
4          Madera                  274.0         -119.635004   

   incident_latitude                           incident_id  \
0          40.774000  2ca11d45-8139-4c16-8af0-880d99b21e82   
1           1.000000  8f61f461-552d-4538-b186-35ab030da416   
2          36.602575  094719ba-a47b-4abb-9ec5-a506b2b9fd23   
3          34.288877  58f89ff8-bd3e-4355-b1c0-8fa05c747d3f   
4          37.116295  357ffc13-bef9-48eb-810f-c5de851972eb   

  incident_date_extinguished  
0       2018-01-09T13:46:00Z  
1       2009-05-25T00:00:00Z  
2       2013-02-28T20:00:00Z  
3       2013-04-22T09:00:00Z  
4       2013-05-01T07:00:00Z

In [13]:
CA_fire_df.shape

(2835, 9)

## California fire Dataset Two

In [15]:
#Import dataset two
CA_fire_df_2 = pd.read_csv('./Data source/California_Fire_Incidents.csv')

In [19]:
CA_fire_df_2.head()

AcresBurned  Active                                          AdminUnit  \
0     257314.0   False  Stanislaus National Forest/Yosemite National Park   
1      30274.0   False  USFS Angeles National Forest/Los Angeles Count...   
2      27531.0   False  CAL FIRE Riverside Unit / San Bernardino Natio...   
3      27440.0   False                              Tahoe National Forest   
4      24251.0   False                       Ventura County Fire/CAL FIRE   

   AirTankers  ArchiveYear  CalFireIncident  \
0         NaN         2013             True   
1         NaN         2013             True   
2         NaN         2013             True   
3         NaN         2013            False   
4         NaN         2013             True   

                            CanonicalUrl  \
0         /incidents/2013/8/17/rim-fire/   
1  /incidents/2013/5/30/powerhouse-fire/   
2    /incidents/2013/7/15/mountain-fire/   
3    /incidents/2013/8/10/american-fire/   
4      /incidents/2013/5/2/springs-fire/   

                                  ConditionStatement ControlStatement  \
0                                                NaN              NaN   
1                                                NaN              NaN   
2                                                NaN              NaN   
3                                                NaN              NaN   
4  Acreage has been reduced based upon more accur...              NaN   

      Counties  ...                                     SearchKeywords  \
0     Tuolumne  ...  Rim Fire, Stanislaus National Forest, Yosemite...   
1  Los Angeles  ...  Powerhouse Fire, May 2013, June 2013, Angeles ...   
2    Riverside  ...  Mountain Fire, July 2013, Highway 243, Highway...   
3       Placer  ...  American Fire, August 2013, Deadwood Ridge, Fo...   
4      Ventura  ...  Springs Fire, May 2013, Highway 101, Camarillo...   

                Started     Status  StructuresDamaged StructuresDestroyed  \
0  2013-08-17T15:25:00Z  Finalized                NaN                 NaN   
1  2013-05-30T15:28:00Z  Finalized                NaN                 NaN   
2  2013-07-15T13:43:00Z  Finalized                NaN                 NaN   
3  2013-08-10T16:30:00Z  Finalized                NaN                 NaN   
4  2013-05-02T07:01:00Z  Finalized                6.0                10.0   

   StructuresEvacuated  StructuresThreatened  \
0                  NaN                   NaN   
1                  NaN                   NaN   
2                  NaN                   NaN   
3                  NaN                   NaN   
4                  NaN                   NaN   

                               UniqueId               Updated  WaterTenders  
0  5fb18d4d-213f-4d83-a179-daaf11939e78  2013-09-06T18:30:00Z           NaN  
1  bf37805e-1cc2-4208-9972-753e47874c87  2013-06-08T18:30:00Z           NaN  
2  a3149fec-4d48-427c-8b2c-59e8b79d59db  2013-07-30T18:00:00Z           NaN  
3  8213f5c7-34fa-403b-a4bc-da2ace6e6625  2013-08-30T08:00:00Z           NaN  
4  46731fb8-3350-4920-bdf7-910ac0eb715c  2013-05-11T06:30:00Z          11.0  

[5 rows x 40 columns]

In [25]:
CA_fire_df_2.columns

Index(['AcresBurned', 'Active', 'AdminUnit', 'AirTankers', 'ArchiveYear',
       'CalFireIncident', 'CanonicalUrl', 'ConditionStatement',
       'ControlStatement', 'Counties', 'CountyIds', 'CrewsInvolved', 'Dozers',
       'Engines', 'Extinguished', 'Fatalities', 'Featured', 'Final',
       'FuelType', 'Helicopters', 'Injuries', 'Latitude', 'Location',
       'Longitude', 'MajorIncident', 'Name', 'PercentContained',
       'PersonnelInvolved', 'Public', 'SearchDescription', 'SearchKeywords',
       'Started', 'Status', 'StructuresDamaged', 'StructuresDestroyed',
       'StructuresEvacuated', 'StructuresThreatened', 'UniqueId', 'Updated',
       'WaterTenders'],
      dtype='object')

In [26]:
#filtered and retained only the relevant columns
CA_fire_df_2 = CA_fire_df_2[['Name','Counties','CrewsInvolved', 'Dozers', 'Engines', 'Fatalities',
                            'Helicopters', 'Injuries','PersonnelInvolved','StructuresDamaged', 'StructuresDestroyed',
                            'StructuresThreatened', 'UniqueId','WaterTenders']]

In [28]:
CA_fire_df_2.shape

(1636, 14)

In [29]:
#Display dataset two
CA_fire_df_2.head()

Name     Counties  CrewsInvolved  Dozers  Engines  Fatalities  \
0         Rim Fire     Tuolumne            NaN     NaN      NaN         NaN   
1  Powerhouse Fire  Los Angeles            NaN     NaN      NaN         NaN   
2    Mountain Fire    Riverside            NaN     NaN      NaN         NaN   
3    American Fire       Placer            NaN     NaN      NaN         NaN   
4     Springs Fire      Ventura           47.0     8.0    117.0         NaN   

   Helicopters  Injuries  PersonnelInvolved  StructuresDamaged  \
0          NaN       NaN                NaN                NaN   
1          NaN       NaN                NaN                NaN   
2          NaN       NaN                NaN                NaN   
3          NaN       NaN                NaN                NaN   
4         11.0      10.0             2167.0                6.0   

   StructuresDestroyed  StructuresThreatened  \
0                  NaN                   NaN   
1                  NaN                   NaN   
2                  NaN                   NaN   
3                  NaN                   NaN   
4                 10.0                   NaN   

                               UniqueId  WaterTenders  
0  5fb18d4d-213f-4d83-a179-daaf11939e78           NaN  
1  bf37805e-1cc2-4208-9972-753e47874c87           NaN  
2  a3149fec-4d48-427c-8b2c-59e8b79d59db           NaN  
3  8213f5c7-34fa-403b-a4bc-da2ace6e6625           NaN  
4  46731fb8-3350-4920-bdf7-910ac0eb715c          11.0

In [34]:
#count number of unique wildfire incident ID
CA_fire_df_2['UniqueId'].nunique()

1609

In [36]:
#Check for dublicated wildfire incidents
CA_fire_df_2[CA_fire_df_2.duplicated(subset=['UniqueId'], keep=False)].head()

Name    Counties  CrewsInvolved  Dozers  \
13                              Grand Fire        Kern           29.0     1.0   
14                              Grand Fire     Ventura           29.0     1.0   
66                                 50 Fire   El Dorado            NaN     NaN   
67                                 50 Fire  Sacramento            NaN     NaN   
148  Sonoma-Lake-Napa Unit Lightning Fires      Sonoma            NaN     NaN   

     Engines  Fatalities  Helicopters  Injuries  PersonnelInvolved  \
13      34.0         NaN          1.0       4.0              847.0   
14      34.0         NaN          1.0       4.0              847.0   
66       NaN         NaN          NaN       NaN                NaN   
67       NaN         NaN          NaN       NaN                NaN   
148      NaN         NaN          NaN       NaN                NaN   

     StructuresDamaged  StructuresDestroyed  StructuresThreatened  \
13                 NaN                  NaN                   NaN   
14                 NaN                  NaN                   NaN   
66                 NaN                  NaN                   NaN   
67                 NaN                  NaN                   NaN   
148                NaN                  NaN                   NaN   

                                 UniqueId  WaterTenders  
13   e0d038e9-0926-4c05-8317-ec7075a04b12          12.0  
14   e0d038e9-0926-4c05-8317-ec7075a04b12          12.0  
66   4630e9ce-88ec-4159-9bc9-25ff5f8216cc           NaN  
67   4630e9ce-88ec-4159-9bc9-25ff5f8216cc           NaN  
148  5ce45b6c-81b5-422c-9078-8c68dd872c8a           NaN

In [38]:
#Remove dublicated wildfire incidents
CA_fire_df_2 = CA_fire_df_2.drop_duplicates(subset=['UniqueId'])

In [40]:
CA_fire_df_2.shape

(1609, 14)

In [42]:
#Select columns with a null value
col_to_fill =['CrewsInvolved', 'Dozers', 'Engines', 'Fatalities', 'Helicopters', 'Injuries',
               'PersonnelInvolved','StructuresDamaged','StructuresDestroyed','WaterTenders','StructuresThreatened']
#Replace null value with 0
CA_fire_df_2[col_to_fill] = CA_fire_df_2[col_to_fill].fillna(0)

In [44]:
CA_fire_df_2.head(5)

Name     Counties  CrewsInvolved  Dozers  Engines  Fatalities  \
0         Rim Fire     Tuolumne            0.0     0.0      0.0         0.0   
1  Powerhouse Fire  Los Angeles            0.0     0.0      0.0         0.0   
2    Mountain Fire    Riverside            0.0     0.0      0.0         0.0   
3    American Fire       Placer            0.0     0.0      0.0         0.0   
4     Springs Fire      Ventura           47.0     8.0    117.0         0.0   

   Helicopters  Injuries  PersonnelInvolved  StructuresDamaged  \
0          0.0       0.0                0.0                0.0   
1          0.0       0.0                0.0                0.0   
2          0.0       0.0                0.0                0.0   
3          0.0       0.0                0.0                0.0   
4         11.0      10.0             2167.0                6.0   

   StructuresDestroyed  StructuresThreatened  \
0                  0.0                   0.0   
1                  0.0                   0.0   
2                  0.0                   0.0   
3                  0.0                   0.0   
4                 10.0                   0.0   

                               UniqueId  WaterTenders  
0  5fb18d4d-213f-4d83-a179-daaf11939e78           0.0  
1  bf37805e-1cc2-4208-9972-753e47874c87           0.0  
2  a3149fec-4d48-427c-8b2c-59e8b79d59db           0.0  
3  8213f5c7-34fa-403b-a4bc-da2ace6e6625           0.0  
4  46731fb8-3350-4920-bdf7-910ac0eb715c          11.0

### Initial Data Import

- Dataset One was imported and filtered to retain only the relevant 9 columns for analysis.
- Dataset Two was imported and filtered to retain only the relevant 14 columns for analysis.

### Handling Duplicates and Missing Data

- Duplicate Removal: Removed duplicate rows based on unique incident IDs in Dataset Two.
- Filling Missing Data: For the following columns in Dataset Two, missing values were filled with 0:
    - CrewsInvolved, Dozers, Engines, Fatalities, Helicopters, Injuries, PersonnelInvolved, StructuresDamaged,
      StructuresDestroyed, WaterTenders, StructuresThreatened.


## Merge the two datasets

In [48]:
#Check for common column between the two datasets
CA_fire_df.columns, CA_fire_df_2.columns

(Index(['incident_name', 'incident_date_created',
        'incident_administrative_unit', 'incident_county',
        'incident_acres_burned', 'incident_longitude', 'incident_latitude',
        'incident_id', 'incident_date_extinguished'],
       dtype='object'),
 Index(['Name', 'Counties', 'CrewsInvolved', 'Dozers', 'Engines', 'Fatalities',
        'Helicopters', 'Injuries', 'PersonnelInvolved', 'StructuresDamaged',
        'StructuresDestroyed', 'StructuresThreatened', 'UniqueId',
        'WaterTenders'],
       dtype='object'))

In [49]:
#renamed the wildfire id column in both datasets to the same name 'Incident ID'
CA_fire_df = CA_fire_df.rename(columns={'incident_id': 'Incident ID'})
CA_fire_df_2 = CA_fire_df_2.rename(columns={'UniqueId': 'Incident ID'})
                               

In [50]:
#Mereged the two datasets in to one dataframe
complete_df = CA_fire_df.merge(CA_fire_df_2, on='Incident ID', how = 'left')

In [51]:
complete_df.shape

(2835, 22)

In [52]:
complete_df.head(5)

incident_name incident_date_created          incident_administrative_unit  \
0    Bridge Fire  2017-10-31T11:22:00Z       Shasta-Trinity National Forest    
1      Pala Fire  2009-05-24T14:56:00Z               CAL FIRE San Diego Unit   
2     River Fire  2013-02-24T08:16:00Z          CAL FIRE San Bernardino Unit   
3  Fawnskin Fire  2013-04-20T17:30:00Z        San Bernardino National Forest   
4      Gold Fire  2013-04-30T12:59:00Z  CAL FIRE Madera-Mariposa-Merced Unit   

  incident_county  incident_acres_burned  incident_longitude  \
0          Shasta                   37.0         -122.309000   
1       San Diego                  122.0            1.000000   
2            Inyo                  407.0         -118.016510   
3  San Bernardino                   30.0         -116.941311   
4          Madera                  274.0         -119.635004   

   incident_latitude                           Incident ID  \
0          40.774000  2ca11d45-8139-4c16-8af0-880d99b21e82   
1           1.000000  8f61f461-552d-4538-b186-35ab030da416   
2          36.602575  094719ba-a47b-4abb-9ec5-a506b2b9fd23   
3          34.288877  58f89ff8-bd3e-4355-b1c0-8fa05c747d3f   
4          37.116295  357ffc13-bef9-48eb-810f-c5de851972eb   

  incident_date_extinguished           Name  ... Dozers  Engines  Fatalities  \
0       2018-01-09T13:46:00Z    Bridge Fire  ...    0.0      0.0         0.0   
1       2009-05-25T00:00:00Z            NaN  ...    NaN      NaN         NaN   
2       2013-02-28T20:00:00Z     River Fire  ...   25.0     25.0         0.0   
3       2013-04-22T09:00:00Z  Fawnskin Fire  ...    0.0      0.0         0.0   
4       2013-05-01T07:00:00Z      Gold Fire  ...    0.0      0.0         0.0   

   Helicopters  Injuries  PersonnelInvolved  StructuresDamaged  \
0          0.0       0.0                0.0                0.0   
1          NaN       NaN                NaN                NaN   
2          0.0       2.0              476.0                0.0   
3          0.0       0.0                0.0                0.0   
4          0.0       0.0                0.0                0.0   

   StructuresDestroyed  StructuresThreatened  WaterTenders  
0                  0.0                   0.0           0.0  
1                  NaN                   NaN           NaN  
2                  0.0                   0.0          25.0  
3                  0.0                   0.0           0.0  
4                  0.0                   0.0           0.0  

[5 rows x 22 columns]

In [53]:
complete_df.columns

Index(['incident_name', 'incident_date_created',
       'incident_administrative_unit', 'incident_county',
       'incident_acres_burned', 'incident_longitude', 'incident_latitude',
       'Incident ID', 'incident_date_extinguished', 'Name', 'Counties',
       'CrewsInvolved', 'Dozers', 'Engines', 'Fatalities', 'Helicopters',
       'Injuries', 'PersonnelInvolved', 'StructuresDamaged',
       'StructuresDestroyed', 'StructuresThreatened', 'WaterTenders'],
      dtype='object')

In [54]:
#removed the repeated columns 
complete_df = complete_df.drop(columns=['Name', 'Counties',])

In [59]:
#Select columns with a null value
col_to_fill =['CrewsInvolved', 'Dozers', 'Engines', 'Fatalities', 'Helicopters', 'Injuries',
               'PersonnelInvolved','StructuresDamaged','StructuresDestroyed','WaterTenders','StructuresThreatened']

#Replace null value with 0
complete_df[col_to_fill] =complete_df[col_to_fill].fillna(0)

In [62]:
complete_df.head(3)

incident_name incident_date_created      incident_administrative_unit  \
0   Bridge Fire  2017-10-31T11:22:00Z   Shasta-Trinity National Forest    
1     Pala Fire  2009-05-24T14:56:00Z           CAL FIRE San Diego Unit   
2    River Fire  2013-02-24T08:16:00Z      CAL FIRE San Bernardino Unit   

  incident_county  incident_acres_burned  incident_longitude  \
0          Shasta                   37.0          -122.30900   
1       San Diego                  122.0             1.00000   
2            Inyo                  407.0          -118.01651   

   incident_latitude                           Incident ID  \
0          40.774000  2ca11d45-8139-4c16-8af0-880d99b21e82   
1           1.000000  8f61f461-552d-4538-b186-35ab030da416   
2          36.602575  094719ba-a47b-4abb-9ec5-a506b2b9fd23   

  incident_date_extinguished  CrewsInvolved  Dozers  Engines  Fatalities  \
0       2018-01-09T13:46:00Z            0.0     0.0      0.0         0.0   
1       2009-05-25T00:00:00Z            0.0     0.0      0.0         0.0   
2       2013-02-28T20:00:00Z           25.0    25.0     25.0         0.0   

   Helicopters  Injuries  PersonnelInvolved  StructuresDamaged  \
0          0.0       0.0                0.0                0.0   
1          0.0       0.0                0.0                0.0   
2          0.0       2.0              476.0                0.0   

   StructuresDestroyed  StructuresThreatened  WaterTenders  
0                  0.0                   0.0           0.0  
1                  0.0                   0.0           0.0  
2                  0.0                   0.0          25.0

In [64]:
#Renamed all the columns
complete_df = complete_df.rename(columns={'incident_name': 'Incident Name',
                                          'incident_date_created' : 'Fire Started',
                                          'incident_administrative_unit': 'Administrative Unit',
                                          'incident_county': 'County',
                                          'incident_acres_burned': 'Acres Burned',
                                          'incident_longitude': 'Longitude', 
                                          'incident_latitude': 'Latitude',
                                          'incident_date_extinguished': 'Fire Extinguished',
                                          'CrewsInvolved': 'Crews Involved',
                                          'PersonnelInvolved': 'Personnel Involved',
                                          'StructuresDamaged': 'Structures Damaged',
                                          'StructuresDestroyed': 'Structures Destroyed',
                                          'StructuresThreatened': 'Structures Threatened', 
                                          'WaterTenders': 'Water Tenders'
                                            })

### Merging Datasets
- The common column between the two datasets was the unique incident ID.
- Datasets were merged on the incident ID column to create a unified dataset.
- For missing rows from 2020 to present, data was filled with 0.
- All columns were renamed for clarity and consistency.

## Missing Values

In [70]:
#Check for null values
complete_df.isnull().sum().sort_values(ascending= False)

Fire Extinguished        625
Administrative Unit       56
Acres Burned              53
County                    10
Incident Name              0
Fatalities                 0
Structures Threatened      0
Structures Destroyed       0
Structures Damaged         0
Personnel Involved         0
Injuries                   0
Helicopters                0
Dozers                     0
Engines                    0
Fire Started               0
Crews Involved             0
Incident ID                0
Latitude                   0
Longitude                  0
Water Tenders              0
dtype: int64

In [72]:
#Check if there is a duplicate in 'Incident ID'
complete_df.duplicated(subset=['Incident ID']).sum()

0

In [74]:
complete_df[complete_df['Acres Burned'].isnull()].head()

Incident Name          Fire Started  \
1440      Oak Fire  2019-10-22T19:20:44Z   
1467   Cashe Fire   2019-11-25T12:02:02Z   
1476    Baker Fire  2020-03-03T15:20:43Z   
1609    Adams Fire  2020-07-27T09:35:49Z   
1620  Cowhead Fire  2020-07-31T20:06:14Z   

                                    Administrative Unit     County  \
1440                   Camp Pendleton Marine Corps Base  San Diego   
1467               Yolo County Fire Protection District       Yolo   
1476                              Tahoe National Forest       Yuba   
1609                        USFS Plumas National Forest     Lassen   
1620  Bureau Land of Management Northern California ...      Modoc   

      Acres Burned   Longitude   Latitude  \
1440           NaN -117.403719  33.351145   
1467           NaN -121.729691  38.734634   
1476           NaN -121.123810  39.454930   
1609           NaN -120.025872  39.922415   
1620           NaN -120.171830  41.522610   

                               Incident ID Fire Extinguished  Crews Involved  \
1440  7264a106-e0f4-41de-8fd0-3f9110431e28               NaN             0.0   
1467  9c26f915-1b33-422d-b30a-9eb4da6fd729               NaN             0.0   
1476  7411452c-1613-4e4f-afe6-865db0c476c4               NaN             0.0   
1609  69aa5821-3027-464a-9c61-daa8b75e090f               NaN             0.0   
1620  5b3fa4b0-d0e0-46d3-9ee8-0b4a9ff7c225               NaN             0.0   

      Dozers  Engines  Fatalities  Helicopters  Injuries  Personnel Involved  \
1440     0.0      0.0         0.0          0.0       0.0                 0.0   
1467     0.0      0.0         0.0          0.0       0.0                 0.0   
1476     0.0      0.0         0.0          0.0       0.0                 0.0   
1609     0.0      0.0         0.0          0.0       0.0                 0.0   
1620     0.0      0.0         0.0          0.0       0.0                 0.0   

      Structures Damaged  Structures Destroyed  Structures Threatened  \
1440                 0.0                   0.0                    0.0   
1467                 0.0                   0.0                    0.0   
1476                 0.0                   0.0                    0.0   
1609                 0.0                   0.0                    0.0   
1620                 0.0                   0.0                    0.0   

      Water Tenders  
1440            0.0  
1467            0.0  
1476            0.0  
1609            0.0  
1620            0.0

In [76]:
#drop rows where 'Acres Burned' is null
complete_df= complete_df.dropna(subset=['Acres Burned'])

In [78]:
complete_df.isnull().mean().sort_values(ascending= False)

Fire Extinguished        0.214234
Administrative Unit      0.019051
County                   0.003595
Incident Name            0.000000
Fatalities               0.000000
Structures Threatened    0.000000
Structures Destroyed     0.000000
Structures Damaged       0.000000
Personnel Involved       0.000000
Injuries                 0.000000
Helicopters              0.000000
Dozers                   0.000000
Engines                  0.000000
Fire Started             0.000000
Crews Involved           0.000000
Incident ID              0.000000
Latitude                 0.000000
Longitude                0.000000
Acres Burned             0.000000
Water Tenders            0.000000
dtype: float64

In [80]:
#check the distribution of 'Acres Burned' to creat a bins
pd.set_option('display.float_format', '{:.2f}'.format)
complete_df['Acres Burned'].describe()

count      2782.00
mean       4271.73
std       35171.79
min           0.00
25%          30.00
50%          83.00
75%         347.00
max     1032648.00
Name: Acres Burned, dtype: float64

### Null values detected:
- Fire Extinguished: 625 missing values
- Administrative Unit: 56 missing values
- Acres Burned: 53 missing values
- County: 10 missing values
  
Handling Null Values

- Duplicate Incident IDs: Checked for and removed duplicates.
- Missing Acres Burned: Rows with missing acres burned data were dropped.

## Filling Missing Values in the 'Fire Extinguished' Column

In [96]:
#make a copy of complete_df 
copy_df = complete_df.copy()

#created a bins for 'Acres Burned' and a new column 'Acres Burnned Bins' to categorize  the acres burned by size
bins = [0, 50, 500, 5000, 50000, 500000, complete_df['Acres Burned'].max()]
copy_df['Acres Burnned Bins'] = pd.cut(copy_df['Acres Burned'], bins= bins)

#Changed 'Fire started' and 'Fire Extinguished' columns data type to datetime
copy_df['Fire Started']= pd.to_datetime(copy_df['Fire Started'])
copy_df['Fire Extinguished']= pd.to_datetime(copy_df['Fire Extinguished'])

#Kept the row with a missing Fire Extinguished date in null_value and droped the rows from copy_df dataframe
null_value = copy_df[copy_df['Fire Extinguished'].isnull()]
copy_df = copy_df.dropna(subset=['Fire Extinguished'])

#In a new column calculated the duration of the wildfire by subtracting the fire started date from fire extinguished date
copy_df.loc[:, 'Days to Extinguish'] = copy_df['Fire Extinguished'] - copy_df['Fire Started']

#Grouped by the bins and calculated the median duration of the wildfire for the respective wildfire size 
median_day = copy_df.groupby('Acres Burnned Bins', observed = False)['Days to Extinguish'].mean()

median_day

Acres Burnned Bins
(0.0, 50.0]             109 days 18:03:27.301712780
(50.0, 500.0]            53 days 23:47:14.151648352
(500.0, 5000.0]          64 days 06:35:59.378461538
(5000.0, 50000.0]           68 days 00:26:51.281250
(50000.0, 500000.0]      77 days 06:44:41.717948718
(500000.0, 1032648.0]              95 days 02:07:32
Name: Days to Extinguish, dtype: timedelta64[ns]

In [159]:
#make a copy of complete_df 
copy_df = complete_df.copy()

#created a bins for 'Acres Burned' and a new column 'Acres Burnned Bins' to categorize  the acres burned by size
bins = [0, 50, 500, 5000, 50000, 500000, complete_df['Acres Burned'].max()]
copy_df['Acres Burnned Bins'] = pd.cut(copy_df['Acres Burned'], bins= bins)

#Changed 'Fire started' and 'Fire Extinguished' columns data type to datetime
copy_df['Fire Started']= pd.to_datetime(copy_df['Fire Started'])
copy_df['Fire Extinguished']= pd.to_datetime(copy_df['Fire Extinguished'])

#Kept the row with a missing Fire Extinguished date in null_value and droped the rows from copy_df dataframe
null_value = copy_df[copy_df['Fire Extinguished'].isnull()]
copy_df = copy_df.dropna(subset=['Fire Extinguished'])

#In a new column calculated the duration of the wildfire by subtracting the fire started date from fire extinguished date
copy_df.loc[:, 'Days to Extinguish'] = copy_df['Fire Extinguished'] - copy_df['Fire Started']
#Grouped by the bins and calculated the median duration of the wildfire for the respective wildfire size 
median_day = copy_df.groupby('Acres Burnned Bins', observed = False)['Days to Extinguish'].median()


#change pandas series to dict
median_day = median_day.to_dict()

#create a column called Median Day and ap the median days to the corresponding bins
null_value['Median Day']= null_value['Acres Burnned Bins'].map(median_day)

#Change column type from category to str then convert to timedelta
null_value['Median Day']= null_value['Median Day'].astype(str)
null_value['Median Day']= pd.to_timedelta(null_value['Median Day'])

#Fill the missing fire extinguished date by adding the median timedelta to the 'Fire Started' column, the drop the median column
null_value['Fire Extinguished'] = null_value['Fire Started'] + null_value['Median Day']
null_value = null_value.drop(columns=['Acres Burnned Bins', 'Median Day'])
#now all the rows with a missing fire Extinguished date will be filled
null_value.head()

Incident Name              Fire Started  \
1270      Far Fire 2019-07-06 20:21:25+00:00   
1271     Lake Fire 2019-07-08 15:30:14+00:00   
1278    Horse Fire 2019-07-12 19:08:54+00:00   
1287  Whiskey Fire 2019-07-16 18:44:15+00:00   
1292   Valley Fire 2019-07-21 12:01:40+00:00   

                       Administrative Unit       County  Acres Burned  \
1270      CAL FIRE Nevada-Yuba-Placer Unit         Yuba         38.00   
1271  CAL FIRE Madera-Mariposa-Merced Unit     Mariposa        244.00   
1278          CAL FIRE Shasta-Trinity Unit       Shasta         30.00   
1287             CAL FIRE Santa Clara Unit  San Joaquin         90.00   
1292  CAL FIRE Madera-Mariposa-Merced Unit     Mariposa        149.00   

      Longitude  Latitude                           Incident ID  \
1270    -121.35     39.06  2d3b577c-7b58-4ee3-9e21-cbec5abd8ce4   
1271    -120.30     37.52  6be8378d-9a89-40a3-bc66-4fc0ab22ebc2   
1278    -122.44     40.46  ab6d9c8e-5fbb-4a0d-a206-30e83ab51877   
1287    -121.47     37.96  12528e0f-03e0-4598-8df0-e3017fa7382a   
1292    -120.03     37.24  f8a2fd4a-286e-4688-8a5c-fa2ad98751ab   

             Fire Extinguished  Crews Involved  Dozers  Engines  Fatalities  \
1270 2019-07-09 10:44:58+00:00            0.00    0.00     0.00        0.00   
1271 2019-07-12 00:27:23+00:00            0.00    0.00     0.00        0.00   
1278 2019-07-15 09:32:27+00:00            0.00    0.00     0.00        0.00   
1287 2019-07-20 03:41:24+00:00            0.00    0.00     0.00        0.00   
1292 2019-07-24 20:58:49+00:00            0.00    0.00     0.00        0.00   

      Helicopters  Injuries  Personnel Involved  Structures Damaged  \
1270         0.00      0.00                0.00                0.00   
1271         0.00      0.00                0.00                0.00   
1278         0.00      0.00                0.00                0.00   
1287         0.00      0.00                0.00                0.00   
1292         0.00      0.00                0.00                0.00   

      Structures Destroyed  Structures Threatened  Water Tenders  
1270                  0.00                   0.00           0.00  
1271                  0.00                   0.00           0.00  
1278                  0.00                   0.00           0.00  
1287                  0.00                   0.00           0.00  
1292                  0.00                   0.00           0.00

In [161]:
#Replace the row with a missing fire Extinguished date in 'Complete_df' with rows from "null_value"
complete_df.loc[complete_df['Fire Extinguished'].isnull(), 'Fire Extinguished'] = null_value['Fire Extinguished']

In [163]:
complete_df.isnull().sum().sort_values(ascending= False)

Administrative Unit      53
County                   10
Incident Name             0
Engines                   0
Structures Threatened     0
Structures Destroyed      0
Structures Damaged        0
Personnel Involved        0
Injuries                  0
Helicopters               0
Fatalities                0
Dozers                    0
Fire Started              0
Crews Involved            0
Fire Extinguished         0
Incident ID               0
Latitude                  0
Longitude                 0
Acres Burned              0
Water Tenders             0
dtype: int64

### Missing Fire Extinguished Date:
- A copy of the original dataset was made for processing.
- Acres Burned was categorized into bins: [0, 50, 500, 5000, 50000, 500000, 500000+].
- The median time to extinguish fires in each bin was calculated and added to the Fire Start Date to estimate the missing Fire Extinguished date.

## Filling Missing Values in 'Administrative Unit' and 'County Unit columns'

In [167]:
complete_df['County'] = complete_df['County'].astype(str)

In [169]:
complete_df['Administrative Unit']= complete_df['Administrative Unit'].fillna('Unknown')

In [171]:
complete_df.iloc[39,3] = 'Shasta'
complete_df.iloc[56,3] = 'Shasta'
complete_df.iloc[236,3] = 'San Joaquin'
complete_df.iloc[313,3] = 'Santa Barbara'
complete_df.iloc[542,3] = 'Imperial'
complete_df.iloc[744,3] = 'Del Norte'
complete_df.iloc[987,3] = 'Alpine'
complete_df.iloc[1051,3] = 'Alpine'
complete_df.iloc[1104,3] = 'Siskiyou'
complete_df.iloc[1719,3] = 'Modoc'

In [173]:
complete_df = complete_df.sort_values('Fire Started')

In [175]:
complete_df.head(5)

Incident Name          Fire Started  \
1124    Bridge Fire  1969-12-31T16:00:00Z   
504     Taglio Fire  1969-12-31T16:00:00Z   
1         Pala Fire  2009-05-24T14:56:00Z   
2        River Fire  2013-02-24T08:16:00Z   
3     Fawnskin Fire  2013-04-20T17:30:00Z   

                       Administrative Unit          County  Acres Burned  \
1124                     Marin County Fire           Marin         45.00   
504   CAL FIRE Madera-Mariposa-Merced Unit          Merced         12.00   
1                  CAL FIRE San Diego Unit       San Diego        122.00   
2             CAL FIRE San Bernardino Unit            Inyo        407.00   
3           San Bernardino National Forest  San Bernardino         30.00   

      Longitude  Latitude                           Incident ID  \
1124    -122.77     38.07  49232e73-c3ba-4dc3-b673-b973a918fa7c   
504     -121.08     37.22  fcb8579e-7469-4c9e-9eac-527264e6f26b   
1          1.00      1.00  8f61f461-552d-4538-b186-35ab030da416   
2       -118.02     36.60  094719ba-a47b-4abb-9ec5-a506b2b9fd23   
3       -116.94     34.29  58f89ff8-bd3e-4355-b1c0-8fa05c747d3f   

         Fire Extinguished  Crews Involved  Dozers  Engines  Fatalities  \
1124  2019-01-04T09:22:00Z            0.00    0.00     0.00        0.00   
504   2018-01-09T10:05:00Z            0.00    0.00     0.00        0.00   
1     2009-05-25T00:00:00Z            0.00    0.00     0.00        0.00   
2     2013-02-28T20:00:00Z           25.00   25.00    25.00        0.00   
3     2013-04-22T09:00:00Z            0.00    0.00     0.00        0.00   

      Helicopters  Injuries  Personnel Involved  Structures Damaged  \
1124         0.00      0.00                0.00                0.00   
504          0.00      0.00                0.00                0.00   
1            0.00      0.00                0.00                0.00   
2            0.00      2.00              476.00                0.00   
3            0.00      0.00                0.00                0.00   

      Structures Destroyed  Structures Threatened  Water Tenders  
1124                  0.00                   0.00           0.00  
504                   0.00                   0.00           0.00  
1                     0.00                   0.00           0.00  
2                     0.00                   0.00          25.00  
3                     0.00                   0.00           0.00

In [177]:
complete_df = complete_df.drop([1124,504,1]) 

In [179]:
complete_df.reset_index(drop=True, inplace=True)

In [113]:
complete_df = complete_df[['Incident Name','County', 'Administrative Unit', 'Acres Burned', 'Crews Involved',
                           'Dozers', 'Engines',	'Fatalities',	'Helicopters',	'Injuries',	'Personnel Involved',
                           'Structures Damaged',	'Structures Destroyed',	'Structures Threatened', 'Water Tenders',
                          'Fire Started','Fire Extinguished','Longitude',	'Latitude', 'Incident ID']]

In [117]:
complete_df.to_csv("./Cleaned Data/california_wildfire_data_cleaned.csv")

### Missing Administrative Unit:
- The 56 missing values were filled with 'Unknown'.
### Missing County: 
- The 10 missing county values were manually filled by searching using longitude and latitude coordinates.
### Export:
- After cleaning, the dataset was successfully exported to the output folder.